### ISSUE TITLE
Heat pump support

### ISSUE THREAD

**Issue by gooooba** [XattSPT/com.melcloud#4]
hi,

please add heatpump support if possible.
OpenHab implementation has pumps already supported.

https://github.com/lucacalcaterra/org.openhab.binding.melcloud

**Comment by gooooba** [comment]
It it helps, some https trace of melcloud working with heatpump(urls, parameters, values, ...)
[melcloud-reverse.txt](https://github.com/XattSPT/com.melcloud/files/3812688/melcloud-reverse.txt)

**Comment by XattSPT** [comment]
It should work with heatpumps.

Try not to set floors, or zones in melcloud app. it’s a bug on my app, I
will try to find time to solve it soon.

I actually use a heatpump, so I’m sure it works 😊

El dc, 6 nov 2019 a les 6:41 gooooba <notifications@github.com> va escriure:

**Comment by gooooba** [comment]
I have default building and Zone 1 setup and no floors.

In homey I can select my building/device(icon is air condition) but it doesnt show any usable option. I see current temperature(null), power(not working), FanPower(not aplicable to heat pump) and mode(with vertical, horisontal, full thermostat) which is not applicable to heat pump.

Plus each time I add a device in homey when I logon to melcloud app I need to agree to conditions and sets my melcloud to spanish.

Please let me know what I'm doing wrong.

**Comment by gooooba** [comment]
I see that there is an update 1.1.2. Tried and still the same regarding heat pump control.

**Comment by XattSPT** [comment]
dear @gooooba, i've been watching links you told me and integrating of Heat Pumps is not easier as I thought. 

Now I'm focused on pulish AC control and adding Flowcards to control them. 

After that I'll try to develop something for Heat Pumps, but it is not easy considering that I do not have a Heat Pump and I'm a hooby time developer, so I'm not sure if I could develop a this app. 

I'll be writting here if I publish a test for Pump Control.

**Comment by gooooba** [comment]
@XattSPT If you would like read only access to my heat pump, I could arrange that. Please let me know if you have time and energy.

**Comment by XattSPT** [comment]
@gooooba please give a try to this... 
   https://homey.app/en-us/app/com.mecloud/MelCloud/test/
and tell me if it works more or less properly, and/or what kind of changes are needed.
If it is a disaster... I probably need those read only access... :)

**Comment by gooooba** [comment]
@XattSPT Thanks:)  It detects the pump. it shows current temperature.

It looks like it doesn't send commands to the pump: I tried On/Off and set the room temperature.
It doesnt show watter tank temperature, just the "-" sign.

The main homey screen tile shows current temp. That is great.

Also cooling and heating changing collors is great idea.

**Comment by XattSPT** [comment]
I should need those read access... to fix those things... :(

El ds, 30 nov 2019 a les 19:00 gooooba <notifications@github.com> va
escriure:

**Comment by XattSPT** [comment]
gerard.ricart@gmail.com

**Comment by XattSPT** [comment]
@gooooba a couple of thing...

   1) I think I'll will need this read access code to test :( (there is my email in las post, after you catch it i'll delete)
  2) I've published a new test version with somes changes... it sould improve sending actions and tankwater temp (I hope).
  3) If I publish on Github, are you able to install on homey? I don't wont to publish all those tries on Homey AppStore... and will be easier for me and you to test and report.

Thanks!

**Comment by gooooba** [comment]
@XattSPT  Got the email, you can delete it.

I installed the latest version. Tank temp shows 47 degrees actual temp is 42, so something must be off...
I don't think the sending part works. Now after changing the temp in homey even after waiting and restart of the app doesnt even pick up back the real set temperature(previous version did read the correct set temp)

I will configure the guest access for you.
You will get an email from melcloud with invitation.

**Comment by gooooba** [comment]
sent you the melcloud request.

I can install from github if push to homey is complicated for you. It is easier for me from homey app :) dont need pc and cmd line for it. But no problem.

**Comment by XattSPT** [comment]
@gooooba great. I received and can see your device.

With this I can check better what's receiving from Melcloud.

I'll priorize homey appstore tests, then.
![9F4E5960-EA4A-4FA2-8C5F-7E0A96F2E54F](https://user-images.githubusercontent.com/38588531/69906358-8ee24c80-13c2-11ea-8589-8410e8095397.png)

**Comment by gooooba** [comment]
@XattSPT  If you could add flow and return flow temperatures while you are working on this would be great. Once you have the hard basics done should be easy :)

**Comment by XattSPT** [comment]
@gooooba give a look on new test on Homey AppStore. I hope we are going to the rigth direcction, your 'read only' access helps me a lot. 

https://homey.app/en-us/app/com.mecloud/MelCloud/test/

Needed to test specially:
1) All commands are working (specially on/off and forcedWater)
2) limitations on target_temperature  are the same as Melcloud (or give me min and max temperatures from app)
3) Added Flows for Forced Water, please try them (Trigger, Condition, Action)

Other questions:
A) It is possible to set Hot and Cold temp on Melcloud App? If true, do you think It's interesting to add also on device settings to set them? I've added them as a sensor, but seem they are settable (called SetHeatFlowTemperatureZone1 and SetCool....) 
B) I can not find hottemp and cooltemp you asked on your post (flow and return). Are those ones mentioned on A)?
B) Witch kind of operable modes can you choose on Melcloud? Do you think it should be added? (I see OperationMode and OperationModeZone1). If yes, I should need operation modes names and correlation to modes (0, 1, 2, ....)
C) I see also a parameter SetTankWaterTemperature (now value 47), do yo think should be settable from device settings?
D) What is EcoHotWater (true or false)? Should I add it to device settings? As a On/off Button?
E) What happens with Zone2? Should I add it? Or add do you think it is better to develop a 2 Zones Device? I don't know how to do it. Your opinion will be welcomed.

thanks in advance!

**Comment by gooooba** [comment]
@XattSPT tried to test and explain as much as possible :)

1. OFF/ON and forceHotWatter works with some cosmetic issues. When I press hotWater button for example the icon blinks briefly like it is ON, but changes back to OFF right away, so it looks like command didn't stick. But after about 30 seconds it changes state to ON like expected. It works like this on all on/off actions(power and watter). Not critical but someone impatient would probably press action again and again as it seems it doesn't hold.

**Please read about the general issue on the bottom of the post.**

2. If I understand correctly you are talking about min/max room temperature. This is just the limitation for guest account from 21 to 22. In my case I can set room temperature from 10 to 30.

3. Tested as much as possible and it looks like the flows work. Except I can't get the forceWatter trigger to work if changed externaly either fromthe pump or melcloud app. If I toggle hot watter button in homey the trigers work for both triggers ON and AUTO.

A) SetHeatFlowTemperatureZone1 and SetCoolFlowTemperatureZone1 is setable via melcloud app.
I only use the SetHeatFlowTemperatureZone1 as I don't have cooling configured via the heat pump in the summer. I use pump only for heating. There are multiple SetHeatFlowTemperatureZone1 parameters in XML and I set the number to 28 degrees for you to find the correct one.
SetHeatFlowTemperatureZone1 parameter is used only when FLOW mode is enabled.
Heat pump has 3 modes. ROOM(thermostat based),FLOW(here the SetHeatFlowTemperatureZone1 is used) and CURVE(this is based on external temperature on predefined logic). I use ROOM mode mostly as I have thermostat in living room. Many people use CURVE mode as they don't have adidtional thermostat while the FLOW mode is fixed watter flow temperature trough the system and is not usefull in real life mostly.

B) Parameters for (flow and return) watter temperature are "FlowTemperature" and "ReturnTemperature". They are displayed when you list devices: https://app.melcloud.com/Mitsubishi.Wifi.Client/User/ListDevices
This are only sensors for the watter flow temperature from heat pump into and out of the system.

B) Operation modes explained in A. corelation is (ROOM=0, FLOW=1,CURVE=2)

C) SetWatterTankTemperature should be settable if possible in settings, not main screen.

D) EcoHotWater is parameter for DHW(domestic hot water) or how the watter tank is heated. This is better suited to be added to settings as it is not used that much but nice to have.

E) I only have one zone, so in my case zone 2 settings are not used. But if someone has 2 zones like two different floors for example, than maybe zone 2 should be almost like a 2 seperate devices. Don't realy know what would be the full impact if 2 zones would be enabled. Some settings like power,forced watter mode, watter tank would be common while the set room temperature and operable modes would be based on the zone. 
Don't know what is possible in homey interface. For me I don't need the zone 2 as it would just bother me :).

Other stuff:
- tank temperature I see now is about 60 degrees. Usualy is max 47 but legionela program was running tonight. This counter the only thing I see on the sensors screen

- please add "OutdoorTemperature", "FlowTemperature" and "ReturnTemperature" parameters under sensor screen

- There is one **general issue.** Any setting I do with homey (power,temp,watter). It sets tank temperature to 0 or 40 degrees depending what is pressed. Power ON/OFF sets tank temperature to 40. forceWatter on sets tank temperature also to 40 and forceWatter OFF sets tank temperature to 0.
Also when the SetHeatFlowTemperatureZone1 paremeter gets set to 25 degrees on any action from homey.

I believe the reason is that you need to always set all parameters not just the one. If I do set forced hot watter from melcloud APP for example it looks like it reads all the current information and then sets all information like it was read, plus the change in this case we want ForcedHotWatterMode parameter to true/false. - this is the URL that changes when setting the forcedWatter for example https://app.melcloud.com/Mitsubishi.Wifi.Client/Device/SetAtw
Try using melcloud on a PC browser and fiddler to check the parameters the app sends. I guess if you set only one parameter then the pump sets the missing parameters to some defaults or min/max values.

You can set the ROOM temperature and check in the melcloud app under unit settings where you can set the flow and tank temperature(you will see that it changes). I gave you more permissions so you can set the tank temp, flow temp and operation mode(please set tank to 47degrees and operation to ROOM when you are finished testing).
SetHeatFlowTemperatureZone1 parameter is only visible in melcloud app if you sellect FLOW operation mode under UNIT SETTINGS.

**Comment by XattSPT** [comment]
@gooooba thanks. I'll take a look this afternoon. I think you deleted all permition, because now I can not see your device on Melcloud.

The general issue you told is probalbly an error on sending commands. I will check carefully.

**Comment by gooooba** [comment]
permissions should be fixed now.

**Comment by XattSPT** [comment]
Thanks @goomba I’ll publish new beta this night with most of your report
solved.

El dg, 1 des 2019 a les 15:27 gooooba <notifications@github.com> va
escriure:

**Comment by XattSPT** [comment]
@gooooba new publish.

I think I solved most of issues you reported. 

Pending to try by your side on/off function, now it should go correctly, but I cannot test by myself.

You need to delelte device and re-add it to view improvements on UI (sensors on last page).

Added also device settings (set temperatures, ecomode, and tanktemp)

Regarding "external" triggering, I need to investigate a little more, you are right and does not work with external changes, something is wrong by my side, I'll will test with my AC devices (happens same issue), and then I will improve this ones.

Next weekend I will try to develop function modes. Be sure I'll not touch your device during weekdays, I've deleted from Homey APP and don't use Melcloud on a daily basis. 

thanks! 

PD: U've earned a mention on Homey app's page as a contributor! Feel free to tell me if you prefer not to appear or you prefer real name instead of nickname in next (and probably public) releases.

**Comment by gooooba** [comment]
@XattSPT, I tested it. It looks like now issue reseting the settings is gone. I can also see all the sensors we talked about. 👍 

**Minor fixes required:**
When you get to it could you please rename sensors:
"HOT WATER TEMP." to "HOT WATER FLOW"
"COLD WATER TEMP." to "COLD WATER FLOW"

Correct min/max values for settings based on what the pump is capable:
Set hot watter tank temperature = min 40 / max 60
Set heat watter zone 1 = min 25 / max 60

**Other:**
- Electric heater status/flows
I would like to monitor when any of the builtin helper electric heaters turns ON. It is a pain to detect it otherway than watch the display on the pump itself day and night :). The heaters themself are controlled by the pump, but status or ability to monitor them or. trigger the flow based on them would be great. This way it would be easier to drill down deep into pump settings and modify timings when the heaters start heating to optimize the energy usage as they are quite power hungry...

This are the parameters for the heaters (true/false):
"BoosterHeater1Status"
"BoosterHeater2Status"
"BoosterHeater2PlusStatus"
"ImmersionHeaterStatus"

In the flows they should be usable as (trigger and condition).
Maybe they could be visible in sensors if it won't be to cluttered with allready existing sensors.
If it is possible to hide/show sensor based on settings in homey it would be great :)

- Is it possible to disable the ability to power OFF the pump from homey device screen. Many times when I try to enter into pump details I don't press long enough or turn it OFF by mistake by brushing the device with the finger. It is a general homey interface issue for me.
Maybe even better than disabling the button, some devices(plugs) in homey have advanced setting where you can set Always On device (yes/no) so they can't be turned off/on via button or flow without going into settings and setting always on parameter to NO.
Heat pump is one of the devices that generaly is always turned ON and don't want to turn it off by mistake.

Thanks for all your hard work. Regarding me as a contributor you decide waht you think is best. I have no problem with any option you mentioned.

**Comment by gooooba** [comment]
@XattSPT You probably allready know, but I noticed there no insights for melcloud device and its paramaters.

**Comment by XattSPT** [comment]
@gooooba 
   -  minor fixes updated.
   - added sensors and Flowcards (trigger and condition). I tried to test them, but i do not want to touch your system more than a ºC or two and I'm not able to get "heaters on", so please try it and verify if they work correctly (sensors and triggers)
   - Added temperatures to insights.
   - Added "always on" on device settings. when activated never stops the pump. I'm not being able to not "turn down" button effect on UI Homey Screen, it goes off and after few seconds it turns on,  never sends off command to device... is all I can achieve at this moment. 
   - Solved external triggering.

I'll try to develop modes tomorrow, I will publish new test then. Writting now to check all your reported is done.

**Comment by XattSPT** [comment]
@gooooba 

do you know the difference between theese ones?

  OperationMode: 2,
  OperationModeZone1: 0,
  OperationModeZone2: 2,

I supose U use OperationModeZone1, but, what is "OperationMode"? Heat, Cold & Auto?

**Comment by XattSPT** [comment]
@gooooba give a try to it...

https://homey.app/en-us/app/com.mecloud/MelCloud/test/

**Comment by gooooba** [comment]
@XattSPT 
I managed to trigger Booster Heater 1, which is used as heating helper.
Tested in a flow (trigger) works OK.
Insights seems to work OK.
OperationMode works OK.
Always powerd on works OK.

Only thing **not** working is Advanced settings. It reads all the values OK. But if I set any or all of them none of them sticks. After next status update the values return back like they were.
Settings: (set watter tank temperature, set heat water zone1, set cool water zone , echo hot water)

I'm will try to trigger imersion heater, but circumstances must be right, so I will let you know if it works.

**Comment by XattSPT** [comment]
@gooooba there is a little bug that U've dicovered and I'm triing to solve... 

just change any parameter (set temperature) after any change on settings and then parameters will be saved!

Working on this!

**Comment by gooooba** [comment]
Tested the heaters some more. ON conditions are working fine but off conditions arent. Also when looking on of history on the last page you can see the ON events for heaters, but OFF events are not visible. The timestamp is there, but no text of what has happened then.
![174D7369-40F5-499E-B741-0B715F0EF3A0](https://user-images.githubusercontent.com/3322016/70368712-9ddb6a00-18ae-11ea-8095-163e2320b117.png)

**Comment by gooooba** [comment]
@XattSPT  Found another interesting sensor for heat pumps named **DefrostMode**
DefrostMode=0 (not defrosting, normal operation)
DefrostMode=2 (defrosting)

And found out what OperationMode sensor means(for some modes):
OperationMode=1 (heating the WatterTank)
OperationMode=2 (heating the house)
OperationMode=5 (nothing, idle)

Probably there other numbers means something else but this is what I could test.

**Comment by XattSPT** [comment]
@gooooba I've published new test version.

Changes on settings... Solved.
Conditions... not sure if I've solved the problem. If you can test them, souldh be helpful.
Defrost ... added as sensor.
Operation Mode ... added as sensor. Also added possiblility to change it on Settings. Added also trigger and condition Flowcards. I've no added Action Flowcard because I think theese actions should not have to be automatic, you opinion will be appreciated.  Mode 4 is defrosting.

Note: Test published do not change operation mode on device (i've blocked it in order to not touch your device and I've forget to unlock :). I'll be unlocking for next version. Elsewhere you can try all functions (trigers, actions, ... or information received from Melcloud changing externally on device or in Melcloud)

**Comment by gooooba** [comment]
@XattSPT will test test test :)

Allready did test the Force hot watter as it is the easiest. When I trigger ForceHotWater from homey I get instant notification from homey as it triggers the flowcard. The Auto card doesn’t trigger when pump changes back to Auto.

I don’t know if I’m reporting back the correct findings as the flowcard for Forced Hot Water triggers every minute now until the ForceHotWater is turned ON on the pump/melcloud. I guess while the pump is in force hot water mode, everytime homey updates from the cloud it triggers the ForceHotWater ON again and again...

I have status update set to 1min, so this is probably reason for 1min triggers.

The status screen shows only once for ForceHotWater ON and once for Auto, but Auto has no text.
As I at first updated the app without deleting the device I noticed that only ForceHotWater Auto is without text. Heaters that were before also without text were after update OK.

Here is a new screenshot after I deleted the device and triggered ForceHotWater:
![D33CC969-C3DD-4BBA-B9AB-20B50EAF84D8](https://user-images.githubusercontent.com/3322016/70388778-85567700-19b6-11ea-943f-5f5a3dca702d.png)

I think ForceHotWater works like this in my case when TankWater is already hot enough becouse of all the testing :)

Before FHW is triggered usualy the pump is in heating mode. The pump still runs normaly from the compresor point but moves the valve from radiators to water tank. The OperationMode is still Heating. Then after a minute or so the pump figures out that the water is allready hot enough and turns the valve back to radiators. The operation mode doesn’t change.

If I set the DHW temp higher so it actualy starts to work then Operation mode changes to DHW like planned. The every minute ForceHotWater ON flowcard keeps triggering until the DHW is heated to desired temperature.

Will test the booster heaters tonight when it will be around 0 and defrosting. They will trigger definitly :)

**Comment by gooooba** [comment]
Here are a notifications from homey while heating the DHW. Notice the DHW-Heating notification two times. It was when DHW was finishing and switching back to Heating operation.

![36A60B31-3A2D-4DF5-BB01-2346D2BF665D](https://user-images.githubusercontent.com/3322016/70389241-044eae00-19bd-11ea-8a55-afbde6fd54e9.png)

**Comment by XattSPT** [comment]
@gooooba new test version in a couple of minuts. 

I think all points solved. Your reports are perfect. 

Flowcard, should go only when status changes, not on every refesh. I think it's solved. 
Regarding 2 notifications on modes, i've done a flowcard on my homey (I cannot changes modes) so I'll will receive notifications similar than yours and I can try to solve the problem (if not solved yet)

**Comment by gooooba** [comment]
@XattSPT It looks better :)

Stil minor issue with trigger for FHW.

When I turn FHW On(homey,cloud), ON Flow is trigered, when pump finishes heating the boiler and the pump notifies homey via the cloud then, ON flow is triggered again.

If I press button(ON) in homey, ON flow is trigered. If I press button(Auto) in homey again then Auto Flow is triggered.

So I guess when homey gets push info from cloud the message is wrong...

Will test other stuf during the week. I checked the weather report and it will be above 5 degrees, so Heater testing will be limited...

**Comment by gooooba** [comment]
@XattSPT Always On device is not working always. Just managed to turn it Off by mistake.
When I turn it off I always get the trigger flow notification, and then in some time(up to a minute) I usualy get Power On notification and the homey icon also changes to ON.

The difference in this case was, that Operationmode was OFF as the room temp was set to 19degrees and the pump had nothing to do. I still got OFF trigger flow, but ON trigger flow never happened and it actually turned OFF. Maybe just a coincidence. It is not a big problem, just reporting :)

Also I don’t realy understand the difference between Standby and Off operation mode. Because to me it looks like the same. Pump is not working but is still turned ON as shown in melcloud and the pump display.

Hopefully having set the pump to 19 last night will help test booster heaters as the pump has some catching up to do.

**Comment by XattSPT** [comment]
@gooooba I've published new test version. I think always con capability is solved. You need homey at least on v.3.0.0.

I'll be finishing translation to spanish this weekend and probably publish the app to store. I'll try to test your last report on FHW.

Please tell me if always on solution works correctly.

**Comment by gooooba** [comment]
@XattSPT I tested FHW and power button and it looks like it is working OK. Only thing I cant find is Power ON/Off flow trigger that was there before.

**Comment by XattSPT** [comment]
@gooooba flow for On/off aren't there because when alwayson is active this capability is deleted and also all flow options. No on/off power is controlled. If you need on/off triggers you should set not always on device with the risk to stop it accidentally.

**Comment by gooooba** [comment]
I noticed  that operation  mode  is not reflected  on the history page.

Also now  trigger for operation mode DHW and Heating  are  just  reverse. Sensor for  operation  mode shows  corect operation.

it is strange like maybe  I need  to uninstall and reinstall  the app again. Will try but  december  is  bussy :)

**Comment by XattSPT** [comment]
@gooooba please check you unit. I think I stoped 5 minutes ago (with a bug on program) and I can not turn it on again... Mode has been set to off. 

Sorry!

**Comment by XattSPT** [comment]
@gooooba I do not know how te put operation mode in history... I'll think about it. 

I've published a new test where operation mode and DHW has been corrected. Thank you very much for your reports. They are very very helpful.

I hope you read last mail... Stoping your HeatPump makes me feel so anxious. :( 
I think it should be a good option to have your coordenates (phone, email,.. ) to notify you quickly if this happen again. (I gave you mail email some time ago, write me there)

I'm soo sorry.

**Comment by gooooba** [comment]
@XattSPT unit is ON and currently heating. It happens when testing :)

**Comment by gooooba** [comment]
@XattSPT now in the latest test version operation mode is wrong in sensors page and in triggers. At least DHW and Heating are reversed.

In advanced settings the operation mode is correct for heating. For DHW i didnt check in settings.

**Comment by gooooba** [comment]
@XattSPT Could you add 'set heat water zone' to flow actions?

![Screenshot_20200111-133657_Homey](https://user-images.githubusercontent.com/3322016/72204431-233ad700-3478-11ea-8086-f9ab9ac30464.jpg)

**Comment by XattSPT** [comment]
Done. Added this Flowcards:
 - Triggers: Hot Water Flow Change, Cold Water Flow Change
 - Conditions: Hot Water is bigger than, Cold Water is lower than
 - Actions: Set Heat Water, Set Cool Water  

Also Tried to solve the problem on trigger mode... 

New Test on: https://homey.app/a/com.mecloud/test/
